Fix #3434: [java] False negatives in AssignmentInOperand rule

UncleOwen opened the pull request:
## Describe the PR

I took @oowekyala's code from #5014, made it apply to the current version of the rule, and fixed all the dogfood issues.
While at it, I fixed some XML issues in AssignmentInOperand.xml

## Related issues

- Fix #3434 
- Closes #5014 

## Ready?

- [X] Added unit tests for fixed bug/feature
- [X] Passing all unit tests
- [X] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [X] Added (in-code) documentation (if needed)

adangel commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AssignmentInOperandRule.java:
```diff
@@ -85,31 +85,31 @@ public Object visit(ASTUnaryExpression node, Object data) {
+        ctx.addViolation(impureExpr);
     }
 
+    @Deprecated
```
```suggestion
    /**
     * @deprecated Since 7.17.0. This method is an implementation detail and will be internalized.
     */
    @Deprecated
```

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/Tokens.java:
```diff
@@ -52,7 +52,11 @@ private void setImage(TokenEntry entry, String newImage) {
 
     private int getImageId(String newImage) {
-        return images.computeIfAbsent(newImage, k -> curImageId++);
+        return images.computeIfAbsent(newImage, k -> {
+            int result = curImageId;
+            curImageId++;
+            return result;
+        });
```
The rule has a property to allow increments/decrements. In our dogfood ruleset, we don't set this yet, but maybe we should? I think, keeping this increment (and others) in a single line, makes it easier to read. What do you think?

The current dogfood ruleset is here:

https://github.com/pmd/build-tools/blob/539357e7298b3201b1f2cc748762c88eb101aafd/src/main/resources/net/sourceforge/pmd/pmd-dogfood-config.xml#L263-L268

adangel reviewed (commented):
Thanks!

The rule implementation looks fine, the regression report also makes sense (with the default configuration of the rule).

I'm unsure about the dogfood issues - whether we should enable for PMD the property "allowIncrementDecrement" of the rule? What do you think?

UncleOwen commented:
> @UncleOwen I've updated build-tools just now with the new property for AssignmentInOperand. Could you update this PR? Most of the dogfood fixes are now not needed anymore (only in AssertUtil and CollectionUtil). Thanks!

Done.

I've also rebased: Since there were unrelated changes to the Rule, I wanted to make sure nothing got garbled during merge.

adangel reviewed (approved):
Thanks, LGTM!
